ignite/pkg/cosmosgen/templates/module/module.ts.tpl

```diff
@@ -0,0 +1,71 @@
+// Generated by Ignite ignite.com/cli
+
+import { StdFee } from "@cosmjs/launchpad";
+import { SigningStargateClient, DeliverTxResponse } from "@cosmjs/stargate";
+import { EncodeObject } from "@cosmjs/proto-signing";
+
+import { Api } from "./rest";
+{{ range .Module.Msgs }}import { {{ .Name }} } from "./types/{{ resolveFile .FilePath }}";
+{{ end }}
+
+{{ range .Module.Msgs }}
+type send{{ .Name }}Params = {
+  value: {{ .Name }},
+  fee?: StdFee,
+  memo?: string
+};
+{{ end }}
+{{ range .Module.Msgs }}
+type {{ camelCase .Name }}Params = {
+  value: {{ .Name }},
+};
+{{ end }}
+
+class Module extends Api<any> {
+	private _client: SigningStargateClient;
+	private _addr: string;
+
+  	constructor(baseUrl: string) {
+		super({
+			baseUrl
+		})
+	}
+
+	public withSigner(client: SigningStargateClient, _addr: string) {
+		this._client = client;
+		this._addr = _addr;
+	}
+
+	public noSigner() {
+		this._client = undefined;
+		this._addr = undefined;
+	}
+
+	{{ range .Module.Msgs }}
+	async send{{ .Name }}({ value, fee, memo }: send{{ .Name }}Params): Promise<DeliverTxResponse> {
+		if (!this._client) {
```

Comment:
This is a run time check, and calling this func without calling `withSigner` first will throw an error.

Users have to read documentation to understand about this concept or they will face with this error on their first try and then go to docs to see what went wrong if they called `sendX` directly.

Instead of letting it even possible, we can remove `sendX` funcs from `Module` so users won't have any way of calling them by mistake. `withSigner` can return a `SignedModule` and there we can have the `sendX` funcs under.

This way we can prevent by mistakes in compile time and use auto complete more efficiently. 

TLDR;

```ts
class Module extends Api<any> {
  	constructor(queryURL: string) {}

	public withSigner(client: SigningStargateClient, txURL: string) SignedModule {}
	
	msgMultiSend({ value }: msgMultiSendParams): EncodeObject { }
	
	msgSend({ value }: msgSendParams): EncodeObject { }
};
```

```ts
class Module extends Api<any> {
	private _client: SigningStargateClient;
	private _addr: string;

	async sendMsgMultiSend({ value, fee, memo }: sendMsgMultiSendParams): Promise<DeliverTxResponse> {}
	
	async sendMsgSend({ value, fee, memo }: sendMsgSendParams): Promise<DeliverTxResponse> {}
};
```

Although this Module API still doesn't feel right to me. Maybe we should create `Query` and `Mutation/Tx` namespaces under `Module` and make them usable independently like we did before.

There is no reason to make the Module API complicated by integrating them into same class.

Reply:
I agree `withSigner` might not be the most obvious, hence room for improvement. 

I don't agree former approach yields simpler DX. 

Pure ts: 
```js
let ignite = createIgnite()
ignite.cosmosBank.withSigner(client) // agreed upon room for improvement
ignite.cosmosBank.sendMsg(payload)
```
Using vue-utils
```js
import { cosmosBank } from 'vue-utils'
let { sendMsg } = cosmosBank()
sendMsg(payload) // already signed-in via `use.ts`
```

Can you share equivalent code using former approach so that we can compare?

Reply:
E.g.: improvement: 

We can just use the [keplr plugin](https://github.com/ignite-hq/web/blob/49875c1184b802bd59121f0779a1479c27357805/packages/plugins/src/keplr.ts#L12) to make the `withSigner` call by consumers unnecessary. Or maybe using the [signer plugin](https://github.com/ignite-hq/web/blob/49875c1184b802bd59121f0779a1479c27357805/packages/plugins/src/signer.ts#L3) is better.

That's the beauty of a simple object being populate via plugins. We can create thin layers without increasing chaos. `Module` can remain untouched and plugins can offer sugar syntax without compromising them (e.g.:`ignite.resetAll()`,  `ignite.createAllMsgs()` , `ignite.signin()`).

I do like `queryClient` and `txClient`. But having them separate/together inside the same file does not change the outmost API that much.

Reply:
Maybe better if set a call for this. I still believe that the default API for the TS client should be as simple as possible without giving any possibility to misuse of the API.

```ts
import { queryClient, txClient } from "generated/modules/bank"

const queryc = queryClient(queryAddr);
const txc = txClient(wallet, txAddress);

await queryc.queryAllBalances(accountAddress);
await txc.singAndBroadcast([txc.msgSend(message)]);
```

I believe any other layers we introduce should be on top of this. Above is a very simple API where there are no unnecessary and complicated dependencies in between.

Reply:
Let's go for a call.

Reply:
```ts
let ignite = createIgnite()
ignite.cosmosBank.withSigner(client) /
```

Another Q, about the design above:

Why do we adding all the modules under `ignite` namespace? What if my project only depends on using one or two modules in the chain and I don't care about others. How do we avoid getting those unused modules in to the final JS bundle? Otherwise, the bundle will have unused code inside which will increase the size for no reason.

Reply:
Q3- `Module`s constructor depends on `baseUrl`, the url of the Query API.

What if I only want to use the TX APIs, why should I be required to initialize the Module with the Query API?

Reply:
> ```ts
> let ignite = createIgnite()
> ignite.cosmosBank.withSigner(client) /
> ```
> 
> Another Q, about the design above:
> 
> Why do we add all the modules under the `ignite` namespace? What if my project only depends on using one or two modules in the chain and I don't care about others.
> 
> How do we avoid getting those unused modules into the final JS bundle? Otherwise, the bundle will have unused code inside which will increase the size of file for no reason.

The overall design prompts pluggable by default. The TS Client is just [an object being populated w/ plugins](https://github.com/ignite-hq/ignite.com/blob/587cd1f516ce003c06af2d2ea20b84be5715afcc/src/generated/tendermint-spn-ts-client/index.ts#L396). Each module is a plugin and exported. 

To create custom clients:
```ts
import { _use } from 'client'

let myClient = _use({
  ...plugCosmosBank(),
  ...plugMyFeature()
})
```

The generated helper fn `create<Repo>` is just a wrapper around `_use `

Reply:
this means the default is all(createIgnite) or nothing(_use)...

Imo, the default should be the minimal setup of router/env/websocket/etc. and no modules

also, use() should be a property of the createXXX() object...

As it stands , you can only do a single _use call...I'd expect somethign more like:

const client = createIgnite();
client.use(module1);
client.use(module2); 

where use returns the augmented client so this is also possible:

client.use(module1).use(module2)

You can also allow:
 client.use([module1,module2]) etc.

Reply:
Yes, I'm afraid if we populate all the modules in the default one, most of the people will just start using it that way.

I would say we could try to come up with a slightly different API to both make the manually approach easy and avoid having unused modules end up in the bundle.